contracts/utils/math/Math.sol

```diff
@@ -220,35 +220,29 @@ library Math {
      * Inspired by Henry S. Warren, Jr.'s "Hacker's Delight" (Chapter 11).
      */
     function sqrt(uint256 a) internal pure returns (uint256) {
-        if (a == 0) {
-            return 0;
-        }
-
-        // For our first guess, we get the biggest power of 2 which is smaller than the square root of the target.
-        //
-        // We know that the "msb" (most significant bit) of our target number `a` is a power of 2 such that we have
-        // `msb(a) <= a < 2*msb(a)`. This value can be written `msb(a)=2**k` with `k=log2(a)`.
-        //
-        // This can be rewritten `2**log2(a) <= a < 2**(log2(a) + 1)`
-        // → `sqrt(2**k) <= sqrt(a) < sqrt(2**(k+1))`
-        // → `2**(k/2) <= sqrt(a) < 2**((k+1)/2) <= 2**(k/2 + 1)`
-        //
-        // Consequently, `2**(log2(a) / 2)` is a good first approximation of `sqrt(a)` with at least 1 correct bit.
-        uint256 result = 1 << (log2(a) >> 1);
-
-        // At this point `result` is an estimation with one bit of precision. We know the true value is a uint128,
-        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
-        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
-        // into the expected uint128 result.
         unchecked {
+            if (a <= 1) { return a; }
+            if (a >= ((1 << 128) - 1)**2) { return (1 << 128) - 1; }
+            uint256 aAux = a;
+            uint256 result = 1;
+            if (aAux >= (1 << 128)) { aAux >>= 128; result <<= 64; }
+            if (aAux >= (1 << 64 )) { aAux >>= 64;  result <<= 32; }
+            if (aAux >= (1 << 32 )) { aAux >>= 32;  result <<= 16; }
+            if (aAux >= (1 << 16 )) { aAux >>= 16;  result <<= 8;  }
+            if (aAux >= (1 << 8  )) { aAux >>= 8;   result <<= 4;  }
+            if (aAux >= (1 << 4  )) { aAux >>= 4;   result <<= 2;  }
+            if (aAux >= (1 << 2  )) {               result <<= 1;  }
+            result += (result >> 1);
             result = (result + a / result) >> 1;
             result = (result + a / result) >> 1;
             result = (result + a / result) >> 1;
             result = (result + a / result) >> 1;
             result = (result + a / result) >> 1;
             result = (result + a / result) >> 1;
```

Comment:
I only count 7 iterations, yet the PDF you linked shows that in some cases, 8 are required. I'm worried about that part. Do you have an example of value that needs 8 iterations that we could run both version of the code with ?

Reply:
First, **Newton iteration** refers to `result = (result + a/result) >> 1`. This is the Newton iteration step when computing square roots via Newton's method; see [here](https://en.wikipedia.org/wiki/Integer_square_root#Algorithm_using_Newton's_method).

So, `result += (result >> 1);` **is not** a Newton iteration but rather part of the initialization. Yes, you could replace it another Newton iteration, but that is more expensive.

As to your comments about "8 cases are required", you are misunderstanding the comment and figure. If **only** Newton iterations are performed **and the same initialization is used that is currently used by OZ, **then** it sometimes takes 8 Newton iterations to reach the correct value. Note that this **does not apply** either in the current OZ code nor the proposed change. That figure was primarily meant to show **why** a good initialization value is required for efficient computation.